### PR TITLE
Add django-ratelimit for rate-limiting of login attempts

### LINKED ISSUE #372: Throttle login api
Login api( /login) can be exploited during DoS attack. The use of rate-limiting technologies can reduce the effects of a denial-of-service (DoS) attacks.

### PR DESCRIPTION
**Pull request by rituranjna**
Add django-ratelimit for rate-limiting of login attempts at the authentication backend level

closes #372.

### CODE REVIEW

**Comment by tapasweni-pathak** [comment]
@rituranjna Thanks for raising the issue and submitting this fix. VMS runs Travis build on every new pull request. You can check the log for this pull request [here](https://travis-ci.org/systers/vms/builds/152274238). There are 40 errors and the build has failed. 

Your next step will be to run these steps locally on your changes and then fix the errors. Travis will show a green tick in such a case.

Please fix the issues present in pull request and update it. 

Let me know if you get stuck while doing any of these.

**Comment by rituranjna** [comment]
@tapasweni-pathak 
I am not sure why selenuin tests are failing with following errors
```
NoSuchElementException: Message: Unable to locate element: {"method":"id","selector":"id_login"}
NoSuchElementException: Message: Unable to locate element: {"method":"link text","selector":"Events"}
NoSuchElementException: Message: Unable to locate element: {"method":"xpath","selector":"//table//tbody//tr[1]/td[5]//a"}
NoSuchElementException: Message: Unable to locate element: {"method":"link text","selector":"Shift Sign Up"}
```

How should i find the root cause of this ?

**Comment by tapasweni-pathak** [comment]
@rituranjna Even I am not able to figure out the exact cause. But I am thinking because you are providing the ```
username':'test@test.com','password':'password' 
```
that may be causing the issue. Basically Selenium is trying to search for those elements and not finding them. 

@smarshy Can you find the issue and help us?

**Comment by rituranjna** [comment]
@tapasweni-pathak The test i have added is passing locally on my laptop , but other  tests  are failing. If these parameters are creating trouble, then this test only should fail ?
Are not  test executions independent  ?

**Comment by tapasweni-pathak** [comment]
>The test i have added is passing locally on my laptop , but other tests are failing. 

@rituranjna I didn't get this. :( Are all test cases passing on your local for this test case?

File: vms/authentication/tests/test_services.py [hunk 1]
```diff
@@ -1,3 +1,14 @@
 from django.test import TestCase
+from authentication.views import login_process
+from django.test.client import RequestFactory
+
 
 # Create your tests here.
+class TestLoginProcess(TestCase):
+
+	def testLoginViewRateLimit(self):
```
Comment (rituranjna) [inline vms/authentication/tests/test_services.py]: @tapasweni-pathak  This test is passing locally.

Comment (tapasweni-pathak) [inline vms/authentication/tests/test_services.py]: @rituranjna Okay. Let's wait for @smarshy's comment.

Comment (smarshy) [inline vms/authentication/tests/test_services.py]: @rituranjna Can you increase the number to 35 or higher? We want to verify something.

Comment (rituranjna) [inline vms/authentication/tests/test_services.py]: @smarshy @tapasweni-pathak 
I updated this PR, but tests are still failing :sob:

**Comment by smarshy** [comment]
@rituranjna And they seem to have increased too. Are you creating this test user in the database?

File: vms/authentication/tests/test_services.py [hunk 2]
```diff
@@ -1,3 +1,19 @@
...
+		request = RequestFactory().post("/login")
+		request.user = User.objects.create_user(
+			username='Test User', email='test@a.com', password='Password')
+		for loginAttemp in range(25):
```
Comment (smarshy) [inline vms/authentication/tests/test_services.py]: Isn't it supposed to be changed here too?

Comment (tapasweni-pathak) [inline vms/authentication/tests/test_services.py]: Isn't this the only place where she has to change it? @rituranjna Did you update the pull request?

File: vms/authentication/views.py [hunk 3]
```diff
@@ -18,3 +21,13 @@ def as_view(request, *args, **kwargs):
         response = func(request, *args, **kwargs)
         return response
     return as_view
+
+
+@ratelimit(key='ip', rate='50/m')
```
Comment (smarshy) [inline vms/authentication/views.py]: @tapasweni-pathak 
She has increased it here in rate

File: vms/authentication/tests/test_services.py (hunk 2 above)
Comment (rituranjna) [inline vms/authentication/tests/test_services.py]: @smarshy @tapasweni-pathak 
I kept this 25 intentionally despite ratelimit being 50/minute.
For every login view access, there is a redirection which hit this view again.
so there are 2 login view access per iteration. so rate limit is hit 50 after 25 iteration.

**Comment by rituranjna** [comment]
@smarshy I was under the impression that objects are rolled back after test completion  automatically?

If my assumption is not true, i will add a tearDown methis in this test class.

**Comment by rituranjna** [comment]
@tapasweni-pathak @smarshy 
i think i have found the issue. All test which will run after my newly added test will fail to login, because ratelimit has already exceeded

**Comment by tapasweni-pathak** [comment]
@rituranjna Yes, that's what we are thinking. 

Great. Can you fix it?

**Comment by rituranjna** [comment]
@tapasweni-pathak 
Is there any way to distinguish between production and development environment ?

**Comment by tapasweni-pathak** [comment]
> Is there any way to distinguish between production and development environment ?

Yes, we have separate branches for production and dev in VMS repository. Contributors needs to fork the devlop branch. Does this answer your question? Let me know if you meant something else.

**Comment by rituranjna** [comment]
@tapasweni-pathak @smarshy 
These tests are failing consistently.
I tried cleaning cache in setup, but it did not work.
If i run  one of these failing  tests  locally, firefox opens but nothings shows up there.

How can i fix these tests ?

File: vms/authentication/tests/test_services.py (hunk 2 above)
Comment (tapasweni-pathak) [inline vms/authentication/tests/test_services.py]: @rituranjna Change this 25 to something more than 35 and update the ratelimit if needed. Let's see then.

Comment (smarshy) [inline vms/authentication/tests/test_services.py]: @rituranjna I think you should try changing this. Which redirection are you referring to?

File: vms/event/tests/test_shiftSignUp.py [hunk 4]
```diff
@@ -32,6 +34,8 @@ def setUpClass(cls):
         super(ShiftSignUp, cls).setUpClass()
 
     def setUp(self):
+        from django.core.cache import cache
+        cache.clear()
```
Comment (smarshy) [inline vms/event/tests/test_shiftSignUp.py]: @rituranjna I am slightly confused :( Why are you clearing the cache here? And why only in this test file? I think there are around 17 test files

**Comment by tapasweni-pathak** [comment]
@rituranjna Were you able to fix this?

**Comment by necessary129** [comment]
> @rituranjna: Is there any way to distinguish between production and development environment?

Well, In `settings.py`, you could add 

    RATELIMIT_ENABLE = True if DEBUG else False

to disable the rate limiting when in `DEBUG` mode or something.

**Comment by smarshy** [comment]
@rituranjna We haven't received any reply from you yet. Were you able to resolve the errors in this PR?

**Comment by rituranjna** [comment]
@smarshy  I am not able to find the root cause.

**Comment by tapasweni-pathak** [comment]
@rituranjna Are you still working on this?

**Comment by tapasweni-pathak** [comment]
Please reopen if you plan to continue to work on this.

**Pull request closed without merging**